service/tbc/crawler.go

```diff
@@ -1215,20 +1220,379 @@ func (s *Server) TxIndexer(ctx context.Context, endHash *chainhash.Hash) error {
+			continue
+		}
+
+		for _, txOut := range tx.MsgTx().TxOut {
+			aPoPTx, err := pop.ParseTransactionL2FromOpReturn(txOut.PkScript)
+			if err != nil {
+				// log.Tracef("error parsing tx l2: %s", err)
+				continue
+			}
+			if _, ok := kssCache[*aPoPTx.L2Keystone.Hash()]; ok {
+				// Multiple keystones may exist in block, only store first
+				continue
+			}
+
+			abvKss := aPoPTx.L2Keystone.Serialize()
+			kssCache[*aPoPTx.L2Keystone.Hash()] = tbcd.Keystone{
+				BlockHash:           *blockHash,
+				AbbreviatedKeystone: abvKss[:],
+			}
+		}
+	}
+	return nil
+}
+
+// indexKeystonesInBlocks indexes txs from the last processed block until the
+// provided end hash, inclusive. It returns the number of blocks processed and
+// the last hash it has processedd.
+func (s *Server) indexKeystonesInBlocks(ctx context.Context, endHash *chainhash.Hash, kss map[chainhash.Hash]tbcd.Keystone) (int, *HashHeight, error) {
+	log.Tracef("indexKeystonesInBlocks")
+	defer log.Tracef("indexKeystonesInBlocks exit")
+
+	// indicates if we have processed endHash and thus have hit the exit
+	// condition.
+	var last *HashHeight
+
+	// Find start hash
+	ksHH, err := s.KeystoneIndexHash(ctx)
+	if err != nil {
+		return 0, last, fmt.Errorf("keystone index hash: %w", err)
+	}
+
+	kssPercentage := 95 // flush cache at >95% capacity
+	blocksProcessed := 0
+	hh := ksHH
+	for {
+		log.Debugf("indexing keystones: %v", hh)
+
+		hash := hh.Hash
+		bh, err := s.db.BlockHeaderByHash(ctx, &hash)
+		if err != nil {
+			return 0, last, fmt.Errorf("block header %v: %w", hash, err)
+		}
+
+		// Index block
+		b, err := s.db.BlockByHash(ctx, &bh.Hash)
+		if err != nil {
+			return 0, last, fmt.Errorf("block by hash %v: %w", bh, err)
+		}
+
+		err = processKeystones(b.Hash(), b.Transactions(), kss)
+		if err != nil {
+			return 0, last, fmt.Errorf("process keystones %v: %w", hh, err)
+		}
+
+		blocksProcessed++
+
+		// Try not to overshoot the cache to prevent costly allocations
+		cp := len(kss) * 100 / s.cfg.MaxCachedKeystones
+		if bh.Height%10000 == 0 || cp > kssPercentage || blocksProcessed == 1 {
+			log.Infof("Tx indexer: %v tx cache %v%%", hh, cp)
+		}
+		if cp > kssPercentage {
+			// Set kssMax to the largest tx capacity seen
+			s.cfg.MaxCachedKeystones = max(len(kss), s.cfg.MaxCachedKeystones)
+			last = hh
+			// Flush
+			break
+		}
+
+		// Exit if we processed the provided end hash
+		if endHash.IsEqual(&hash) {
+			last = hh
+			break
+		}
+
+		// Move to next block
+		height := bh.Height + 1
+		bhs, err := s.db.BlockHeadersByHeight(ctx, height)
+		if err != nil {
+			return 0, last, fmt.Errorf("block headers by height %v: %w",
+				height, err)
+		}
+		index, err := s.findPathFromHash(ctx, endHash, bhs)
+		if err != nil {
+			return 0, last, fmt.Errorf("could not determine canonical path %v: %w",
+				height, err)
+		}
+		// Verify it connects to parent
+		if !hash.IsEqual(bhs[index].ParentHash()) {
+			return 0, last, fmt.Errorf("%v does not connect to: %v",
```

Comment:
hey @marcopeereboom what is the intention here? If a block doesn't connect to the expected parent (I assume a re-org could cause this?) then we just stop indexing?

Reply:
it means we do not have all data and thus we cannot continue. Either we have data corruption or a blockheader was missed (which can happen in the insane world of testnet3). We return to give the caller the opportunity to reinsert the blockheader and block. @max-sanchez actually does that in the embedded tbc service in op-geth.